File: src/knowledge/services/providers/local_embedding.py
"""
Local Embedding Provider 实现

支持本地 Embedding 模型，包括 FastEmbed 和 Sentence-Transformers。
"""

import asyncio
import time

from ..models import (
    EmbeddingResponse,
    ProviderType,
    TokenUsage,
    EmbeddingError,
)


# 模型维度映射
MODEL_DIMENSIONS: dict[str, int] = {
    "BAAI/bge-small-zh-v1.5": 512,
    "BAAI/bge-base-zh-v1.5": 768,
    "BAAI/bge-large-zh-v1.5": 1024,
    "BAAI/bge-small-en-v1.5": 384,
    "BAAI/bge-base-en-v1.5": 768,
    "BAAI/bge-large-en-v1.5": 1024,
    "sentence-transformers/all-MiniLM-L6-v2": 384,
    "sentence-transformers/all-mpnet-base-v2": 768,
}


class LocalEmbeddingProvider:
    """Local Embedding Provider 实现

    支持 FastEmbed 和 Sentence-Transformers 后端。
    """

    def __init__(
        self,
        model_name: str = "BAAI/bge-small-zh-v1.5",
        backend: str = "fastembed",
    ) -> None:
        """初始化 Local Embedding Provider

        Args:
            model_name: 模型名称
            backend: 后端类型 ('fastembed' 或 'sentence-transformers')
        """
        self._model_name = model_name
        self._backend = backend
        self._model = None

    @property
    def provider_type(self) -> ProviderType:
        return ProviderType.LOCAL

    def get_dimensions(self, model: str) -> int:
        if model in MODEL_DIMENSIONS:
            return MODEL_DIMENSIONS[model]
        for key, dim in MODEL_DIMENSIONS.items():
            if model in key or key in model:
                return dim
        return 768

    async def _ensure_model(self) -> None:
        """确保模型已加载"""
        if self._model is not None:
            return

        loop = asyncio.get_event_loop()

        if self._backend == "fastembed":
            try:
                from fastembed import TextEmbedding
            except ImportError as e:
                raise EmbeddingError(
                    "fastembed package not installed. Run: pip install fastembed",
                    provider=ProviderType.LOCAL,
                ) from e
            self._model = await loop.run_in_executor(
                None, lambda: TextEmbedding(model_name=self._model_name)
            )
        elif self._backend == "sentence-transformers":
            try:
                from sentence_transformers import SentenceTransformer
            except ImportError as e:
                raise EmbeddingError(
                    "sentence-transformers package not installed. Run: pip install sentence-transformers",
                    provider=ProviderType.LOCAL,
                ) from e
            self._model = await loop.run_in_executor(
                None, lambda: SentenceTransformer(self._model_name)
            )
        else:
            raise EmbeddingError(
                f"Unknown backend: {self._backend}",
                provider=ProviderType.LOCAL,
            )

    async def embed(
        self,
        texts: list[str],
        model: str,
        *,
        dimensions: int | None = None,
    ) -> EmbeddingResponse:
        """执行向量编码请求"""
        await self._ensure_model()
        start_time = time.perf_counter()
        loop = asyncio.get_event_loop()

        try:
            if self._backend == "fastembed":
                embeddings_gen = await loop.run_in_executor(
                    None, lambda: list(self._model.embed(texts))
                )
                embeddings = [emb.tolist() for emb in embeddings_gen]
            else:
                embeddings_np = await loop.run_in_executor(
                    None, lambda: self._model.encode(texts)
                )
                embeddings = embeddings_np.tolist()
        except Exception as e:
            raise EmbeddingError(str(e), provider=ProviderType.LOCAL) from e

        latency_ms = int((time.perf_counter() - start_time) * 1000)
        actual_dimensions = len(embeddings[0]) if embeddings else 0

        return EmbeddingResponse(
            embeddings=embeddings,
            model_used=model or self._model_name,
            provider=ProviderType.LOCAL,
            dimensions=actual_dimensions,
            usage=TokenUsage(),
            latency_ms=latency_ms,
        )

    async def health_check(self) -> bool:
        try:
            await self.embed(texts=["test"], model=self._model_name)
            return True
        except Exception:
            return False

File: src-ts/knowledge/providers/local-embedding.ts
/**
 * Knowledge module - Local Embedding Provider
 *
 * Supports local embedding models including FastEmbed and Sentence-Transformers.
 * In the TypeScript context, this uses dynamic imports for optional backends.
 */

import type { EmbeddingResponse } from '../models';
import {
  ProviderType,
  createTokenUsage,
  createEmbeddingResponse,
  EmbeddingError,
} from '../models';

// ============================================================
// Model dimensions mapping
// ============================================================

const MODEL_DIMENSIONS: Record<string, number> = {
  'BAAI/bge-small-zh-v1.5': 512,
  'BAAI/bge-base-zh-v1.5': 768,
  'BAAI/bge-large-zh-v1.5': 1024,
  'BAAI/bge-small-en-v1.5': 384,
  'BAAI/bge-base-en-v1.5': 768,
  'BAAI/bge-large-en-v1.5': 1024,
  'sentence-transformers/all-MiniLM-L6-v2': 384,
  'sentence-transformers/all-mpnet-base-v2': 768,
};

// ============================================================
// Local Embedding Provider
// ============================================================

export class LocalEmbeddingProvider {
  private readonly _modelName: string;
  private readonly _backend: string;
  private _model: unknown = null;

  constructor(params: {
    modelName?: string;
    backend?: string;
  }) {
    this._modelName = params.modelName ?? 'BAAI/bge-small-zh-v1.5';
    this._backend = params.backend ?? 'fastembed';
  }

  get providerType(): string {
    return ProviderType.LOCAL;
  }

  getDimensions(model: string): number {
    if (model in MODEL_DIMENSIONS) {
      return MODEL_DIMENSIONS[model];
    }
    for (const [key, dim] of Object.entries(MODEL_DIMENSIONS)) {
      if (model.includes(key) || key.includes(model)) {
        return dim;
      }
    }
    return 768;
  }

  /**
   * Ensure model is loaded
   *
   * Attempts to dynamically import the backend module.
   * In Node.js environment, this will use the fastembed or sentence-transformers
   * package if available. Falls back to a stub if not installed.
   */
  private async _ensureModel(): Promise<void> {
    if (this._model !== null) return;

    if (this._backend === 'fastembed') {
      try {
        // Dynamic import for optional dependency
        const mod = await import('fastembed');
        const FlagEmbedding = mod.FlagEmbedding;
        const EmbeddingModel = mod.EmbeddingModel;
        if (!FlagEmbedding) {
          throw new EmbeddingError(
            'fastembed package does not export FlagEmbedding',
            ProviderType.LOCAL,
          );
        }
        // Map model name to EmbeddingModel enum value
        const modelMapping: Record<string, string> = {
          'BAAI/bge-small-zh-v1.5': EmbeddingModel.BGESmallZH,
          'BAAI/bge-small-en-v1.5': EmbeddingModel.BGESmallENV15,
          'BAAI/bge-base-en-v1.5': EmbeddingModel.BGEBaseENV15,
          'BAAI/bge-base-en': EmbeddingModel.BGEBaseEN,
          'BAAI/bge-small-en': EmbeddingModel.BGESmallEN,
          'sentence-transformers/all-MiniLM-L6-v2': EmbeddingModel.AllMiniLML6V2,
        };
        const modelKey = modelMapping[this._modelName] ?? EmbeddingModel.BGESmallZH;
        this._model = await FlagEmbedding.init({ model: modelKey as any });
      } catch (e) {
        if (e instanceof EmbeddingError) throw e;
        throw new EmbeddingError(
          'fastembed package not installed. Run: npm install fastembed',
          ProviderType.LOCAL,
        );
      }
    } else if (this._backend === 'sentence-transformers') {
      try {
        // Dynamic import for optional dependency
        const mod = await import('@xenova/transformers') as any;
        const pipeline = mod.pipeline;
        if (!pipeline) {
          throw new EmbeddingError(
            '@xenova/transformers package does not export pipeline',
            ProviderType.LOCAL,
          );
        }
        this._model = await pipeline('feature-extraction', this._modelName);
      } catch (e) {
        if (e instanceof EmbeddingError) throw e;
        throw new EmbeddingError(
          '@xenova/transformers package not installed. Run: npm install @xenova/transformers',
          ProviderType.LOCAL,
        );
      }
    } else {
      throw new EmbeddingError(
        `Unknown backend: ${this._backend}`,
        ProviderType.LOCAL,
      );
    }
  }

  /**
   * Execute vector encoding request
   */
  async embed(
    texts: string[],
    model: string,
    options?: { dimensions?: number },
  ): Promise<EmbeddingResponse> {
    await this._ensureModel();
    const startTime = performance.now();

    let embeddings: number[][];
    try {
      if (this._backend === 'fastembed') {
        const embedFn = (this._model as any).embed?.bind(this._model);
        if (!embedFn) {
          throw new EmbeddingError('Model does not have embed method', ProviderType.LOCAL);
        }
        const result = await embedFn(texts);
        // fastembed returns async generator or array of typed arrays
        if (Symbol.asyncIterator in result) {
          const chunks: number[][] = [];
          for await (const chunk of result) {
            chunks.push(Array.from(chunk));
          }
          embeddings = chunks;
        } else {
          embeddings = (result as any[]).map((e: any) => Array.from(e));
        }
      } else {
        // sentence-transformers / xenova/transformers
        const encodeFn = (this._model as any);
        const result = await encodeFn(texts, { pooling: 'mean', normalize: true });
        // result is a tensor
        if (result.tolist) {
          embeddings = result.tolist();
        } else if (Array.isArray(result)) {
          embeddings = result;
        } else {
          embeddings = Array.from(result.data || result);
        }
      }
    } catch (e) {
      if (e instanceof EmbeddingError) throw e;
      throw new EmbeddingError((e as Error).message, ProviderType.LOCAL);
    }

    const latencyMs = Math.round(performance.now() - startTime);
    const actualDimensions = embeddings.length > 0 ? embeddings[0].length : 0;

    return createEmbeddingResponse({
      embeddings,
      modelUsed: model || this._modelName,
      provider: ProviderType.LOCAL,
      dimensions: actualDimensions,
      usage: createTokenUsage(),
      latencyMs,
    });
  }

  /**
   * Check provider health status
   */
  async healthCheck(): Promise<boolean> {
    try {
      await this.embed(['test'], this._modelName);
      return true;
    } catch {
      return false;
    }
  }
}
